### ISSUE TITLE
hug needs more documentation

### ISSUE THREAD

**Issue by timothycrosley** [timothycrosley/hug#319]
hug has a collection of documentations on hug.rest and on the github repo (including examples etc) which has served as a reasonable starting point. However, for continued growth, a wider depth of and a higher quality of documentation is an absolute necessity and it is the number 1 priority for the project. 

Feel free to add any ideas for additional documentation sources that could be useful for the project to this issue. The currently known ones are:

- FAQ
- Known Issues / Gotchas
- Best Practices
- Example starting points
- Full documentation for all exposed APIs and features

I, and the entire community, would be grateful for any help in this area.

Thanks!

~Timothy

**Comment by timothycrosley** [comment]
As a precursor for this work as recommended by @vortec, I'll be working on the following items: 

- A single format for documentation (markdown, sphinx, etc TBD)
- A single source for documentation (currently some documentation lives on hug_website and some lives in the core hug repo - I'll be working on moving it all to a single location TBD)
- A table of context showing everything that needs to be documented
- A set of documentation guidelines to help direct how things should be documented

**Comment by mcgrue** [comment]
I recommend putting some simple sample string interpolation on routes...

```
@hug.get('/happy_birthday/{name}')
def happy_birthday(name, age:hug.types.number=1):
    """Says happy birthday to a user"""
    return "Happy {age} Birthday {name}!".format(**locals())
```

...would have saved me some time, and is easy to understand.

**Comment by Kentoseth** [comment]
@timothycrosley  

> "A single format for documentation (markdown, sphinx, etc TBD)"
 > "A single source for documentation"

Why don't we use a single documentation-format that lives on both readthedocs.io and the Hug website (the website can just be a duplicate of readthedocs). However, I have never written or used readthedocs myself, so if it is difficult to use, I would suggest Markdown as the alternative instead (that can then just live on the website)

 > A table of context showing everything that needs to be documented
 > A set of documentation guidelines to help direct how things should be documented

Will you be interested in experimenting with GitHub Projects? Creating a project with "ToDo, Assigned, Done" might be useful for managing all the things required to be done to improve Hug.

**Comment by blaisep** [comment]
@timothycrosley My counsel would be to think of the documentation as a Sphinx project (which could be hosted in readthdocs or locally). I base this suggestion on the following:
    * Sphinx is the _de facto_ doc compiler for Python projects.
    * You don't have to use rST. That said, Markdown is intended for projects that can be described in a single page. rST is intended for multipart projects where you wish to follow the `DRY` principle. You don't have to make it as complex as, say, Djangoproject.org.
   * You're more likely to get ad hoc assistance from people who are into building useful docs (like me).

**Comment by butla** [comment]
I agree with @blaisep that Sphinx and rST are the way to go. Though rST syntax is a bit strange for my taste, it's great for code documentation.

And yes, it would also make me contribute to docs website more (I've seen some unclear descriptions there). Is it generated from something right now? Because the HTML (especially for code snippets in "routing") looks machine-generated.

**Comment by philipaconrad** [comment]
I'm game to help with this docs issue, particularly getting some sort of reference documentation up.

I've been bitten numerous times by the distinct lack of reference docs for Hug's classes/APIs, and in some cases, I've had to dive into the `/examples` folder, just to figure out how to do something.

Ideally, something of similar form to the docs for [Falcon][falcon] or [requests][requests] would be ideal. Both projects have a blend of informal learning resources and examples, along with heavy, in-depth reference docs for the developers who need it.

   [falcon]: https://falcon.readthedocs.io/en/stable/
   [requests]: http://docs.python-requests.org/en/master/

**Comment by aldanor** [comment]
Agreed, as a newcomer, hug is really hard to get around. I had to search information in pull requests, code, closed issues and examples just to figure how things work in general... (e.g. creating API objects)

Re: rST -- there's always an option to use something markdown-based like `gitbook`, although then you'll lose the autodoc capabilities of Sphinx and other Python-related niceties.

**Comment by radhikasundararaman24** [comment]
Hello: I'd like to help the community with some doc support. Looks like this thread was opened long time back, are you still looking for documentation support?

**Comment by piccolbo** [comment]
May I recommend: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/ One can write docstrings in google or numpy formats (which are markdown based, with some structure) and get them straight into readthedocs. I am not sure that the web docs need to reproduce readthedocs. The tutorials can be on one site and the reference on the other.

**Comment by Annakan** [comment]
Given how old this issue is It might not be relevant to the problem any more but the problem is still a big one for Hug.

Let it be said that I do think Hug is a great tool and a potential game changer in the Python API/REST framework world but right now it is really hard to do something non trivial that is not cargo-cult copy from the few examples on the website.

First the website has a bug that does not display code formatting on the [first learn page] (http://www.hug.rest/website/learn/). This page is often the one you get from a google search so it gives a bad impression (like this page is not read often otherwise this would have been caught since it works everywhere else)

Second, i was instantly puzzled in the tutorial by the `@hug.local()` that is explained nowhere i looked for, the referring sentence  is " with local access only" that lead me to think it has to do with security (nope at all) and the the demo-gif shows me a "import mycode, mycode(xxx,xxx)" that looks trivial until you realize that some injection is happening with the third parameter.

What about a definition of what `@hug.local()` does, even grossly done: `@hug.local()` allows you to mark the function for usage a like normal python code through import while benefiting of type checking of your API type declaration, dependency injection like `hug_timer=3` and many other Hug features.

Third, most of the narrative documentation is for Hug 1.0 and the juicy features of 2.0 are hidden on the '2.4.8' link on the top left (not even the current hug version, and nothing tells you it is a link to documentation) that is mostly a a teaser for the new stuff : 
  
 * marshmallow support .... ?
 * class based routing ? : there are two examples one with a endpoint and a mysterious require parameter, the other without them, none of them are serious classes since they don't carry any state and thus sidestep real usage pattern.
* etc ...

I am really convinced that Hug is a very good, well conceived framework with a clever abstraction level and I would really love to use it but I see it for now as a black hole of uncertainty about how to do non trivial things, and that is scary( I scare easy ;) ).

To give some context I just implemented a very simple Rest API with Falcon that I found being rough, low level, but well documented and in the end very usable.

I decided to try to see how much Hug would simplify the code and hit a wall as soon as a wanted to decorate my class : I can't pass to Hug a factory or instanciate the API class myself before handing it to HUG and thus don't know how to make the same simple API work. 
To explain: the REST API needs access to the underlying engine that manage the resource exposed though REST so I need to pass that engine to the API class ( __init__  being the natural point), but HUG takes control of initialisation and so reduce the class to a bunch of stateless functions)
